### PR TITLE
Add support z-loss in pre-training

### PR DESCRIPTION
**Pull request by gagika**
# Description

This PR implements Z-loss to improve numerical stability and prevent runaway logits. Alongside techniques like QK-normalization and logit soft-capping, it is a key mechanism for stabilizing low-precision (BF16/FP8) training.

**Key Changes:**
- **Configuration:** Added a `z_loss_multiplier` parameter to `types.py` (defaults to `0.0`).
- **Integration:** Wired the existing Z-loss utility (`max_utils.cross_entropy_with_logits`) into the standard training loop (`loss_fn`) and the vocabulary tiling path (`vocab_tiling_linen_loss`).
- **Logging:** Normalized Z-loss is now exported to the `aux` dictionary and logged to TensorBoard as `learning/z_loss`.

# Tests

- **Math Verification:** Added `test_cross_entropy_with_z_loss` in `max_utils_test.py` to verify the penalty calculation is mathematically correct.
- **Integration/Tiling:** Added `test_vocab_tiling_gradient_with_z_loss` in `tiling_test.py` to ensure loss and gradients match exactly between standard and vocabulary-tiled computations when Z-loss is enabled.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -462,6 +462,7 @@ class Logits(BaseModel):
       None,
       description="Soft-cap value for the final logits. None or 0.0 means no cap.",
   )
+  z_loss_multiplier: float = Field(0.0, description="The multiplier for the z-loss (e.g., 1e-4). 0.0 to disable.")
```
Comment (aireenmei) [inline src/maxtext/configs/types.py:465]: Do we need to add it to base.yml?

File: src/maxtext/trainers/pre_train/train.py
Comment (aireenmei) [inline src/maxtext/trainers/pre_train/train.py:207]: As the Gemini also pointed out, total_loss is normalized to total_weights. Should z_loss be normalized the same way? Otherwise there seems to be logging inconsistency:
  "learning/loss": loss,          # = total_loss / total_weights  (per-token)
  "learning/z_loss": z_loss,      # = total_z_loss  (unnormalized sum)

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py:465]: +1 It’s not functionally necessary, but it adds clarity.

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -343,6 +363,7 @@ def move(path, value):
 
   scalar_metrics = {
       "learning/loss": loss,
+      "learning/z_loss": z_loss,
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/trainers/pre_train/train.py:371]: Also, wondering if you could see this metric in TB for sanity check?

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (gagika) [reply src/maxtext/configs/types.py:465]: done

File: src/maxtext/trainers/pre_train/train.py
Comment (gagika) [reply src/maxtext/trainers/pre_train/train.py:207]: done

File: src/maxtext/trainers/pre_train/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gagika) [reply src/maxtext/trainers/pre_train/train.py:371]: tested today, it shows in TB: https://screenshot.googleplex.com/7WDjUT6nfer2tC2

File: src/maxtext/trainers/pre_train/train.py [hunk 3]
```diff
@@ -164,11 +176,17 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
...
+
     # Mask out paddings at the end of each example.
     xent = xent * (data["targets_segmentation"] != 0)
+    z_loss = z_loss * (data["targets_segmentation"] != 0)
+
     total_loss = jnp.sum(xent)
+    total_z_loss = jnp.sum(z_loss)
 
   total_weights = jnp.sum(data["targets_segmentation"] != 0)
   # If gradient accumulation is enabled, we don't need to divide total_loss
```
Comment (RissyRan) [inline src/maxtext/trainers/pre_train/train.py:192]: Sorry I missed this part earlier. Shall we also handle gradient accumulation? 

If not, we could have an assertion in type.py that mentioning that `gradient_accumulation_steps` and `z_loss_multiplier` cannot be set together.


https://github.com/AI-Hypercomputer/maxtext/blob/a3475bf2d5048d94fde8dcaa634a5d64e20de113/src/maxtext/trainers/pre_train/train.py#L200-L207

Comment (gagika) [reply src/maxtext/trainers/pre_train/train.py:192]: `loss` already has `z_loss` portion it it (`loss += total_z_loss` inside `cross_entropy_with_logits`, and will be used for gradient / backwards pass.

The standalone `total_z_loss` is only used for metric logging to TensorBoard. Because it doesn't affect the gradients, we don't need to worry about the gradient accumulation  division optimization e.g. we just normalize it by total_weights so the logged metric is always a consistent per-token average.
